src/GovUk.Education.ExploreEducationStatistics.Notifier/GovUk.Education.ExploreEducationStatistics.Notifier.csproj

```diff
@@ -14,6 +14,11 @@
     <PackageReference Include="System.IdentityModel.Tokens.Jwt" Version="6.32.1" />
     <PackageReference Include="Microsoft.IdentityModel.Tokens" Version="6.32.1" />
   </ItemGroup>
+  <ItemGroup>
+    <FunctionsPreservedDependencies Include="Microsoft.IdentityModel.Logging.dll" />
+    <FunctionsPreservedDependencies Include="Microsoft.IdentityModel.Tokens.dll" />
+    <FunctionsPreservedDependencies Include="System.IdentityModel.Tokens.Jwt.dll" />
```

Comment:
Yes that's correct, the Functions runtime host provides a set of standard class libraries. They will be at a different version depending on the Functions host version. 

The latest Function core tools version that's downloading by default in Rider (4.0.5198) is using Azure Function runtime  host 4.21.1 and from doing a bit of debug I can see these provided libraries are at version 6.24.0.

The build process is comparing files in the build output directory against a list of provided libraries by filename (not including version) and removing them to produce a smaller build output.  You can see this in https://github.com/Azure/azure-functions-host/blob/b42df4d4e63ab87f7a737f5a65dadf3eb7ba7de4/tools/ExtensionsMetadataGenerator/src/ExtensionsMetadataGenerator/BuildTasks/RemoveRuntimeDependencies.cs#L36.

In Azure I don't think we any control over what the target 4.x version is so I don't think we can bump anything up to get more up-to-date-versions.

The options seem to be

- Downgrade back to versions provided.
- Disable cleaning during the build entirely with a build property `<_FunctionsSkipCleanOutput>true</_FunctionsSkipCleanOutput>` so that all dependencies are output. This results in a much bigger build though.
- Selectively preserve dependencies like I did here, using `FunctionsPreservedDependencies` items.

If you do either of the latter and in code load the `Assembly` by name, and print the location you should see them being referenced from the path of the project deployment, rather than the path of the function host.

```
    [FunctionName("Test")]
    public static async Task<IActionResult> RunAsync(
        [HttpTrigger(AuthorizationLevel.Function, "get", Route = null)] HttpRequest req)
    {
        var names = new List<string>
        {
            "Microsoft.IdentityModel.Tokens",
        };

        var results = names.Select(assemblyName =>
        {
            var assembly = Assembly.Load(assemblyName);
            var location = assembly.Location;
            var version = assembly.GetName().Version;
            return new AssemblyInfo(assemblyName, location, version);
        }).ToList();

        return new OkObjectResult(results);
    }

    private record AssemblyInfo(string Name, string Location, Version Version);
```

Reply:
Cool, makes sense, approved!